framework/source/class/qx/core/ObjectRegistry.js

```diff
@@ -162,15 +125,57 @@ qx.Bootstrap.define("qx.core.ObjectRegistry",
       }
 
       // Create new hash code
-      var cache = this.__freeHashes;
-      if (cache.length > 0) {
-        hash = cache.pop();
-      } else {
-        hash = (this.__nextHash++) + this.__postId;
-      }
+      hash = this.createHashCode();
 
       // Store
-      return obj.$$hash = hash;
+      obj.$$hash = hash;
+
+      if (qx.core.Environment.get("qx.debug.dispose")) {
+        if (qx.dev && qx.dev.Debug && qx.dev.Debug.disposeProfilingActive) {
+          this.__stackTraces[hash] = qx.dev.StackTrace.getStackTrace();
+        }
+      }
+      
+      return obj.$$hash;
+    },
+    
+    
+    /**
+     * Creates a hash code; this will be a UUID v4, unless the environment variable 
+     * qx.hashCode.classic is set to true
+     * 
+     * @return {String}
+     */
+    createHashCode: function() {
+      if (qx.core.Environment.get("qx.hashCode.classic")) {
+        return String((this.__nextHash++) + this.__postId);
+      } else {
+        /*
+         * Creates a UUID v4; this is taken from https://gist.github.com/jed/982883
+         */
+        function createUuidV4(a) {
+          return a           // if the placeholder was passed, return
+            ? (              // a random number from 0 to 15
+              a ^            // unless b is 8,
+              Math.random()  // in which case
+              * 16           // a random number from
+              >> a/4         // 8 to 11
+              ).toString(16) // in hexadecimal
+            : (              // or otherwise a concatenated string:
+              [1e7] +        // 10000000 +
+              -1e3 +         // -1000 +
+              -4e3 +         // -4000 +
+              -8e3 +         // -80000000 +
+              -1e11          // -100000000000,
+              ).replace(     // replacing
+                /[018]/g,    // zeroes, ones, and eights with
+                createUuidV4 // random hex digits
+              )
+        }
```

Comment:
Please use the readable version of this. We don't care about eliminating a few bytes as that gist is attempting; we do very much care about readability. I had to read half way through that long reference gist discussion to figure out what the heck they were optimizing.

```
function()
{
  return (
    "00000000-0000-4000-8000-000000000000"
      .replace(
        /0/g,
        function()
          {
            return(0 | Math.random() * 16).toString(16)
          }))
}
```